pkg/instancegroups/scaleup.go

```diff
@@ -0,0 +1,196 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instancegroups
+
+import (
+	"context"
+	"fmt"
+
+	api "k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/cloudinstances"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// scaleProvider session data
+type scaleProvider struct {
+	GroupUpdate *RollingUpdateInstanceGroup
+}
+
+// NewScaleUpRollout creates and returns a scale provider
+func NewScaleUpRollout(update *RollingUpdateInstanceGroup) Rollout {
+	return &scaleProvider{GroupUpdate: update}
+}
+
+// RollingUpdate performs a duplicate rolling update of an instance group
+func (p *scaleProvider) RollingUpdate(ctx context.Context, list *api.InstanceGroupList) error {
+	group := p.GroupUpdate.CloudGroup.InstanceGroup
+	fullname := p.GroupUpdate.Name
+	name := group.Name
+	strategy := group.Spec.Strategy
+	update := p.GroupUpdate.Update
+
+	// @step: increase the size of the instancegroup by x percent
+	var newMinSize, newMaxSize, oldMinSize, oldMaxSize int32
+	total := int32(p.GroupUpdate.CloudGroup.Size())
+	oldMinSize = int32(p.GroupUpdate.CloudGroup.MinSize)
+	oldMaxSize = int32(p.GroupUpdate.CloudGroup.MaxSize)
+	newMinSize = int32(p.GroupUpdate.CloudGroup.Size() * 2)
+	newMaxSize = int32(p.GroupUpdate.CloudGroup.MaxSize)
+	if newMinSize > newMaxSize {
+		newMaxSize = newMinSize
+	}
+	update.Infof("strategy: %s, adjusting size setting on instancegroup: %s, min/max (%d/%d)",
+		strategy.Name, name, newMinSize, newMaxSize)
+
+	update.Infof("updating the configuration with adjusted instancegroup settings for: %s", name)
+
+	// @step: grab the configuration from source
+	config, err := update.Clientset.InstanceGroupsFor(update.Cluster).Get(name, v1.GetOptions{})
+	if err != nil {
+		return update.Errorf("unable to retrieve configuration for instancegroup: %s, error: %v", name, err)
+	}
+	config.Spec.MinSize = &newMinSize
+	config.Spec.MaxSize = &newMaxSize
+
+	// @step: attempt to adjust the size of the instancegroup
+	if _, err := update.Clientset.InstanceGroupsFor(update.Cluster).Update(config); err != nil {
+		return update.Errorf("unable to update instancegroup: %s configuration, error: %v", name, err)
+	}
+
+	// @step: we need to update the termination policy to ensure if there is an error the new nodes are the
+	// are the ones which get kicked out first
+	update.Infof("updating termination policy on the instancegroup: %s to newest first", name)
+	if err := update.Cloud.SetTerminationPolicy(update.Cluster, fullname, cloudinstances.TerminateNewestPolicy); err != nil {
```

Comment:
Adding a few keys points .. So i've added a new [SetTerminationPolicy](https://github.com/kubernetes/kops/pull/4038/files#diff-31083c19e09ee7bbc1a3b47f8ef9520bR41) in the fi.Cloud interface. The reasoning behind this, assuming a rollout using scale-up; it doubles and validates but lets say fails on the validation. Without controlling the termination policy of the ASG your stuck here, you can scale down because it could take out the old nodes, thus we update with the Newest and before scale-down we revert to the Oldest. Note, I don't know enough about GCE to mimic this behavior so i've simply turned the strategy off for all bar AWS